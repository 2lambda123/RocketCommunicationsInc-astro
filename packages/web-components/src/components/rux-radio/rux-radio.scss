@use '../../../node_modules/@astrouxds/design-tokens/dist/scss/_colors-dark.scss' as *;
@use '../../../node_modules/@astrouxds/design-tokens/dist/scss/variables.scss' as *;
:host {
    *,
    *::before,
    *::after {
        box-sizing: inherit;
    }

    display: inline-block;
    box-sizing: border-box;
}

.rux-form-field {
    display: flex;
    flex-direction: column;
    font-family: var(--font-body-1-font-family, $font-body-1-font-family);
    font-size: var(--font-body-1-font-size, $font-body-1-font-size);
    letter-spacing: var(
        --font-body-1-letter-spacing,
        $font-body-1-letter-spacing
    );
    font-weight: var(--font-body-1-font-weight, $font-body-1-font-weight);
    color: var(--color-text-primary, $color-text-primary);
}

.rux-radio {
    display: flex;
    position: relative;
    line-height: 1.2;

    input[type='radio'] {
        -webkit-appearance: none;
        appearance: none;

        + label {
            position: relative;
            display: flex;
            align-items: center;
            justify-content: flex-start;
            margin-left: -8px;
            color: var(--color-text-primary, $color-text-primary);
            letter-spacing: 0.5px;
            cursor: pointer;

            &::before {
                box-sizing: border-box;
                display: flex;
                flex-shrink: 0;
                flex-grow: 0;
                content: '';
                align-self: start;
                height: 1.125rem;
                width: 1.125rem;
                margin: 0 0.625rem 0 0;
                border: 1px solid
                    var(
                        --color-border-interactive-muted,
                        $color-border-interactive-muted
                    );
                border-radius: var(--radius-circle, $radius-circle);
                background-color: var(
                    --color-background-base-default,
                    $color-background-base-default
                );
            }

            &::after {
                position: absolute;
                top: 5px;
                display: flex;
                content: '';
            }
        }
        //If no label is passed in we don't want the margin
        + .rux-radio--no-label {
            &::before {
                margin: 0;
            }
        }

        &:checked {
            + label {
                &::before {
                    background-color: var(
                        --color-background-base-default,
                        $color-background-base-default
                    );
                    border-color: var(
                        --color-border-interactive-muted,
                        $color-border-interactive-muted
                    );
                }

                &::after {
                    position: absolute;
                    top: 5px;
                    display: flex;
                    content: '';
                    left: 5px;
                    height: 8px;
                    width: 8px;
                    border-radius: var(--radius-circle, $radius-circle);

                    /* box-shadow: inset 0 0 1px 0 rgba(255, 255, 255, 0.9); */

                    background-color: var(
                        --color-background-interactive-default
                    );
                }
            }
        }

        &:disabled {
            + label {
                cursor: not-allowed;
                opacity: var(--disabled-opacity);
            }
        }

        &:not(:disabled) {
            &:hover,
            &:checked:hover {
                + label {
                    &::before {
                        border-color: var(
                            --color-background-interactive-hover,
                            $color-background-interactive-hover
                        );
                    }
                    &::after {
                        background-color: var(
                            --color-background-interactive-default
                        );
                    }
                }
            }
        }
    }
}

:host([hidden]) {
    display: none;
}
