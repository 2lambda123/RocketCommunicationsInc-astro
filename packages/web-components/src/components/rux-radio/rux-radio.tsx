import {
    Component,
    h,
    Prop,
    Element,
    Event,
    EventEmitter,
    State,
    Watch,
} from '@stencil/core'
import { hasSlot } from '../../utils/utils'

let id = 0

/**
 * @slot (default) - The radio label
 * @part form-field - the form field of the radio
 * @part label - the label of the radio
 */

@Component({
    tag: 'rux-radio',
    styleUrl: 'rux-radio.scss',
    shadow: true,
})
export class RuxRadio {
    private radioId = `rux-radio-${++id}`
    private radioGroup: HTMLRuxRadioGroupElement | null = null

    @Element() el!: HTMLRuxRadioElement

    @State() hasLabelSlot = false

    /**
     * The radio name
     */
    @Prop() name = ''
    /**
     * The radio value
     */
    @Prop() value: string = ''

    /**
     * Toggles checked state of a radio
     */
    @Prop({ reflect: true, mutable: true }) checked: boolean = false

    /**
     * Disables the radio via HTML disabled attribute. Radio takes on a distinct visual state. Cursor uses the not-allowed system replacement and all keyboard and mouse events are ignored.
     */
    @Prop({ reflect: true }) disabled: boolean = false

    /**
     * The radio label text. For HTML content, use the default slot instead.
     */
    @Prop() label?: string
    @Watch('label')
    handleLabelChange() {
        this._handleSlotChange()
    }
    /**
     * Fired when an element has lost focus - [HTMLElement/blur_event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)
     */
    @Event({ eventName: 'ruxblur' }) ruxBlur!: EventEmitter

    connectedCallback() {
        this._onChange = this._onChange.bind(this)
        this._handleSlotChange = this._handleSlotChange.bind(this)
        this.radioGroup = this.el.closest('rux-radio-group')
        this._syncFromGroup = this._syncFromGroup.bind(this)
        if (this.radioGroup) {
            this._syncFromGroup()
            this.radioGroup.addEventListener('ruxchange', this._syncFromGroup)
        }
    }

    componentWillLoad() {
        this._handleSlotChange()
    }

    disconnectedCallback() {
        if (this.radioGroup) {
            this.radioGroup.removeEventListener(
                'ruxchange',
                this._syncFromGroup
            )
        }
    }

    private _handleSlotChange() {
        this.hasLabelSlot = hasSlot(this.el)
    }

    /**
     * Sets checked property when the parent Radio Group value changes.
     */
    private _syncFromGroup() {
        if (this.radioGroup && this.radioGroup.value) {
            this.checked = this.radioGroup.value === this.value
        }
    }

    private _onChange(e: Event): void {
        const target = e.target as HTMLInputElement
        this.checked = target.checked
    }

    private _onBlur = () => {
        this.ruxBlur.emit()
    }

    get hasLabel() {
        return this.label ? true : this.hasLabelSlot
    }

    render() {
        const {
            label,
            radioId,
            checked,
            disabled,
            name,
            value,
            _onChange,
            _onBlur,
            hasLabel,
        } = this

        let part = hasLabel ? 'field' : 'field --no-label'
        let icon = checked ? 'radio-button-checked' : 'radio-button-unchecked'

        return (
            <div part={part}>
                <input
                    type="radio"
                    name={name}
                    id={radioId}
                    disabled={disabled}
                    checked={checked}
                    value={value}
                    onChange={_onChange}
                    onBlur={_onBlur}
                />
                <rux-icon icon={icon} size="18px" part="indicator" />
                <label part="label" htmlFor={radioId}>
                    <slot>{label}</slot>
                </label>
            </div>
        )
    }
}
