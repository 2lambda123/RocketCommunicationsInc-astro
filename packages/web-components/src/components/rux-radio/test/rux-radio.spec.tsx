import { newSpecPage } from '@stencil/core/testing'
import { RuxRadio } from '../rux-radio'

describe('rux-radio', () => {
    it('builds', async () => {
        const radio = new RuxRadio()
        const { radioId, checked, disabled, name, value } = radio

        expect(radio).toBeTruthy()
        expect({
            radioId,
            checked,
            disabled,
            name,
            value,
        }).toEqual({
            radioId: 'rux-radio-1',
            checked: false,
            disabled: false,
            name: '',
            value: '',
        })
    })

    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxRadio],
            html: `<rux-radio value="test"></rux-radio>`,
        })
        expect(page.root).toMatchSnapshot()
    })
})
