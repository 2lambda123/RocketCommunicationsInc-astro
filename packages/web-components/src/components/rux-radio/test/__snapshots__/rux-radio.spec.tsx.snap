// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-radio renders 1`] = `
<rux-radio value="test">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <div class="rux-radio">
        <input id="rux-radio-2" type="radio" value="test">
        <label class="rux-radio--no-label" htmlfor="rux-radio-2" part="label">
          <slot></slot>
        </label>
      </div>
    </div>
  </mock:shadow-root>
</rux-radio>
`;
